File: Python/003-longestSubstring.py
# 91.36 135ms
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start = maxLen = 0
        appeard = {}
        for i in range(len(s)):
            if s[i] in appeard and start <= appeard[s[i]]:
                start = appeard[s[i]] + 1
            else:
                maxLen = max(maxLen, i - start +1)

            appeard[s[i]] = i
        return maxLen

File: Golang/003-longestSubstring.go
// 46.94%  哎
func lengthOfLongestSubstring(s string) int {
	hash := map[byte]int{}
	start := 0
	max := 0
	for i := range s {
		_, ok := hash[s[i]]

		if ok && start <= hash[s[i]] {
			start = hash[s[i]] + 1
		} else {
			if max < i-start+1 {
				max = i - start + 1
			}
		}
		hash[s[i]] = i

	}
	return max
}
